File: code/backend/app/api/sites.py
"""
园区列表 API 路由 (Step 14 + 新建园区)。

2 个接口:
  GET  /sites                  列出当前租户所有 site (读, demo 可用)
  POST /sites                  新建命名园区 (写, demo 拦)

为什么独立成 sites.py 而不是放 query.py:
  query.py 的路由前缀是 /query, 如果把 /sites 放进去会变成
  /api/v1/query/sites, 跟其他模块的 /api/v1/sites/{id}/scene (visual.py)
  路径前缀不一致。统一 /api/v1/sites/* 前缀, 后续 /sites/{id}/buildings
  等接口都放这里。

GET 只读用 get_current_user; POST 写操作挂 require_write_access 拦 demo。
"""
from fastapi import APIRouter, Depends, HTTPException, status

from app.core.deps import CurrentUser, get_current_user, require_write_access
from app.core.response import error, success
from app.models.site import SiteCreateRequest
from app.services.site_service import create_site, list_sites


router = APIRouter(prefix="/sites", tags=["sites"])


@router.get("")
def get_sites(user: CurrentUser = Depends(get_current_user)):
    """列出当前租户所有 site。

    返回 [{site_id, site_code, site_name, latitude, longitude, building_count}]。
    demo 用户可用 (只读)。
    """
    sites = list_sites(user.tenant_id)
    return success(data=sites)


@router.post("", status_code=201)
def post_site(
    req: SiteCreateRequest,
    user: CurrentUser = Depends(require_write_access),
):
    """新建命名园区。

    body:
        {"site_name": "我的园区", "site_code": null, "timezone": "UTC"}

    site_code 空则自动生成。返 {site_id, site_code, site_name}。
    """
    try:
        result = create_site(user.tenant_id, req.site_name, req.site_code, req.timezone)
    except ValueError as ex:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error(message=str(ex), code=400),
        )
    return success(data=result, message="园区已创建")

File: code/frontend/src/api/sites.ts
// ============================================================================
// Sites API - 园区列表 + 园区数据时间范围 (自动检测)
// ----------------------------------------------------------------------------
// 顶栏站点切换器的数据源。DefaultLayout onMounted 调 listSites() 拿到所有
// site, 填进下拉框。context.setSite(site_id) 触发 Park.vue 重载 scene。
//
// getSiteDataRange 给前端用来初始化默认时间范围: demo BDG2 数据是 2017 全年,
// 默认 '本月' (2026-07) 会查到全空, 自动切到数据实际范围才能看到图。
// ============================================================================

import { api } from './client'

export interface SiteItem {
  site_id: string
  site_code: string
  site_name: string
  latitude: number | null
  longitude: number | null
  building_count: number
}

export interface SiteDataRange {
  site_id: string
  // ISO 字符串 (形如 2017-01-01T00:00:00+00:00), 没数据时为 null
  earliest_ts: string | null
  latest_ts: string | null
  // 该园区下的 point_reading 总条数, 0 表示完全没数据
  point_count: number
}

export interface SiteCreateResult {
  site_id: string
  site_code: string
  site_name: string
}

export const sitesApi = {
  listSites() {
    return api.get<SiteItem[]>('/sites')
  },

  /** 新建命名园区。site_code 空则后端自动生成。 */
  createSite(payload: { site_name: string; site_code?: string | null; timezone?: string }) {
    return api.post<SiteCreateResult>('/sites', payload)
  },

  // 走 /query/sites/{id}/data-range 而不是 /sites/{id}/data-range, 因为:
  //   1. query router 已经有 park overview / buildings 等 site-scoped 读查询
  //   2. /sites 路由主要是写操作 (创建/编辑 site), data-range 是只读聚合查询
  getSiteDataRange(siteId: string) {
    return api.get<SiteDataRange>(`/query/sites/${siteId}/data-range`)
  },
}
